dvci/shclinks
# Add endpoints for redirect method

# Summary
This PR establishes the initial framework for the SMART Health Links Prototype - Redirect Method.

For more information, consult the following documentation:
* Josh’s prototype app: https://github.com/jmandel/shclinks
* Redirect method HackMD documentation: https://hackmd.io/qVo4lZB8SDy73rTPsWhKpQ
* ExpressJS: https://expressjs.com/

This prototype implements the most basic redirect method, and so it does *not* contain any encryption, flags, expiration dates, etc.

## New behavior
Assuming the Data Sharer has uploaded files of interest, POSTing to the `/qr` endpoint will ask the resource server to generate a QR Link for a Sharing Manifest for a Sharing Manifest or individual file. The JSON QR payload that is produced currently consists only of a `url` for the authorization server.

The Data Recipient can “redeem” a link by sending a `POST` request to the .url` property from the JSON payload. If the request is valid, the server returns a 301 Moved Permanently status and redirects the client to where the file(s) is/are exposed.

## Code changes
* Establish environment variables
* Create logger for logging server endpoints
* Creation of server.js, which creates an express app and establishes various endpoints for the redirect method

# Testing guidance
It is recommended to test using the [Insomnia API Client](https://insomnia.rest/) or [Postman](https://www.postman.com/).

Reproduce the following:
1. Start the server with `npm start`
2. Send a POST request to http://localhost:3000/qr with the following request body:

```json
{
	"access": [
		{
			"type": "shclink-view",
			"locations": [
				"http://localhost:3000/example1.json",
				"http://localhost:3000/example2.json"
			],
			"datatypes": ["application/fhir+json"]
		}
	]
}
```

You should receive a 200 response status, and the response body should be a JSON object containing a `url` of the form` http://localhost:3000/qr/{random id}/claim`.

3. Send a GET request to the `url` that you received in the previous step. (GET `http://localhost:3000/qr/{random id}/claim`). You should receive a 200 response status with a JSON object containing a `locations` attribute. These locations are for the initial files that were referenced in the original request body.
4. Send a GET request to one of the locations that you received. (GET `http://localhost:3000/qr/{random id}/claimed/{claimed id}/files/{file id}`).
You should receive the content of the file that was referenced in the original request body. (So, for example, if you specified "http://localhost:3000/example1.json" as a location and then follow the endpoint to retrieve that file, you should retrieve the test example1 file in the test directory of this repo).
